ibis/pandas/aggcontext.py

```diff
@@ -364,52 +366,54 @@ def _window_agg_udf(
     This is because pandas's rolling function doesn't support
     multi param UDFs.
     """
-
-    def create_input_iter(
-        grouped_series: SeriesGroupBy, window_size: int
-    ) -> Iterator[np.ndarray]:
-        # create a generator for each input series
-        # the generator will yield a slice of the
-        # input series for each valid window
-        data = getattr(grouped_series, 'obj', grouped_series).values
-        window_size_array = window_size.values
-        for i in range(len(window_size_array)):
-            k = window_size.index[i]
-            yield data[k - window_size_array[i] + 1 : k + 1]
-
     obj = getattr(grouped_data, 'obj', grouped_data)
 
+    assert len(window_lower_indices) == len(window_upper_indices)
+    assert len(window_lower_indices) == len(mask)
+
     # Compute window indices and manually roll
     # over the window.
     # If an window has only nan values, we output nan for
     # the window result. This follows pandas rolling apply
     # behavior.
-    raw_window_size = windowed.apply(len, raw=True).reset_index(drop=True)
-    mask = ~(raw_window_size.isna())
-    window_size = raw_window_size[mask].astype('i8')
-    window_size_array = window_size.values
-
     # If there is no args, then the UDF only takes a single
     # input which is defined by grouped_data
     # This is a complication due to the lack of standard
     # way to pass multiple input pd.Series/SeriesGroupBy
     # to AggregationContext.agg()
     inputs = args if len(args) > 0 else [grouped_data]
 
+    masked_window_lower_indices = window_lower_indices[mask].astype('i8')
+    masked_window_upper_indices = window_upper_indices[mask].astype('i8')
+
+    def create_input_iter(
```

Comment:
can you make this module level and just pass the masks?

Reply:
Extracted to `create_window_input_iter` at the module level